### PR TITLE
Read openmc meshtally

### PR DESCRIPTION
**Pull request by zxkjack123**
This PR adds functions to support reading meshtally data and created corresponding `MeshTally` instance from OpenMC state point file.
Major changes:
- Functions to read mesh, flux and other data required from state point file are added
- An function to create `MeshTally` from given meshtally data is added
- Test functions for read openmc meshtally are added
- Check tag_names are moved to `mesh.py` as a check for `tag_names.setter`.

### CODE REVIEW

File: pyne/mesh.py [hunk 1]
```diff
@@ -1797,3 +1816,25 @@ def _cell_fracs_sort_vol_frac_reverse(cell_fracs):
...
+        raise ValueError("The given tag_names is not an Iterable.")
+
+    # check length of 4
+    if len(tag_names) != 4:
+        raise ValueError("The length of tag_names is not 4.")
+
+    # check content strings
+    for item in tag_names:
+        if not isinstance(item, str):
+            raise ValueError("The content of tag_names ",
```
Comment (pshriwise) [inline pyne/mesh.py]:
```suggestion
            raise ValueError("The content of tag_names "
```

File: pyne/openmc_utils.py [hunk 2]
```diff
@@ -156,3 +171,270 @@ def xml(self):
...
+                                 [z_bounds1, z_bounds2, ...]]
+    """
+
+    sp = openmc.StatePoint(filename)
+    tally = sp.get_tally(id=tally_id)
+    # check filters to find MeshFilter
+    for flt in tally.filters:
+        if isinstance(flt, openmc.filter.MeshFilter):
+            mesh_filter = flt
+    structured_coords = calc_structured_coords(
```
Comment (pshriwise) [inline pyne/openmc_utils.py:235]: We should probably check that the `MeshFilter.mesh` is an `openmc.RegularMesh` class here, otherwise some of the attributes below might not be available. Additionally, I don’t think it would be much work to go ahead an support the `openmc.RectilinearMesh` class, but that could also be added in a future PR.

File: pyne/openmc_utils.py [hunk 3]
```diff
@@ -156,3 +171,270 @@ def xml(self):
...
+    res_tot = np.divide(res_tot, ve_vol)
+    # In the result of OpenMC, x changes fastest
+    # change the order to z changes fastest
+    result = result_changes_order(result, m.dims[3:6])
+    rel_err = result_changes_order(rel_err, m.dims[3:6])
+    res_tot = result_changes_order(res_tot, m.dims[3:6])
+    rel_err_tot = result_changes_order(rel_err_tot, m.dims[3:6])
+    return result, rel_err, res_tot, rel_err_tot
+
+def result_changes_order(result, dims):
```
Comment (pshriwise) [inline pyne/openmc_utils.py:348]: I think the entirety of this function can be accomplished with:

```python
results = results[::-1]
```

File: pyne/openmc_utils.py [hunk 4]
```diff
@@ -156,3 +171,270 @@ def xml(self):
...
+    # assign particle
+    if particle != None:
+        m.particle = particle
+    # assign tag_names
+    if tag_names is None:
+        m.set_default_tag_names()
+    else:
+        m.tag_names = tag_names
+    # check tally_num exist
+    structured_coords = get_structured_coords_from_openmc_sp(
```
Comment (pshriwise) [inline pyne/openmc_utils.py:430]:
```suggestion
m.x_bounds, m.y_bounds, m.z_bounds = 
    structured_coords = get_structured_coords_from_openmc_sp(
```

File: pyne/openmc_utils.py [hunk 5]
```diff
@@ -156,3 +171,270 @@ def xml(self):
...
+        filename, tally_id=m.tally_number)
+
+    # parameters to create mesh
+    m.x_bounds = structured_coords[0]
```
Comment (pshriwise) [inline pyne/openmc_utils.py]: Remove these three lines if the above suggestion is applied.

File: pyne/openmc_utils.py [hunk 6]
```diff
@@ -156,3 +171,270 @@ def xml(self):
...
+    m.y_bounds = structured_coords[1]
+    m.z_bounds = structured_coords[2]
+    m.dims = [len(m.x_bounds) - 1, len(m.y_bounds) - 1, len(m.z_bounds) - 1]
```
Comment (pshriwise) [inline pyne/openmc_utils.py:435]: Probably want to make this a `tuple` for the same reasons we talked about regarding `x/y/z_bounds`.

File: pyne/openmc_utils.py [hunk 7]
```diff
@@ -156,3 +171,270 @@ def xml(self):
...
+    m.num_ves = (len(m.x_bounds)-1) * (len(m.y_bounds)-1) * (len(m.z_bounds)-1)
+    m.e_bounds = get_e_bounds_from_openmc_sp(filename, m.tally_number)
+    m.num_e_groups = len(m.e_bounds) - 1
+    mats = () if mesh_has_mats is True else None
```
Comment (pshriwise) [inline pyne/openmc_utils.py]: So what’s goin on here?

File: pyne/mesh.py [hunk 8]
```diff
@@ -1563,13 +1565,30 @@ def __init__(self):
 
         self.tally_number = None
         self.particle = 'neutron'
-        self.tag_names = None
+        self.set_default_tag_names()
+
+    @property
```
Comment (pshriwise) [inline pyne/mesh.py:1570]: Nice update!

File: pyne/openmc_utils.py [hunk 9]
```diff
@@ -156,3 +171,270 @@ def xml(self):
...
+        objects.
+
+    Returns:
+    --------
+    m : MeshTally object
+        The MeshTally object created from OpenMC state point file with tally
+        number of tally_id.
+    """
+
+    m = MeshTally()
```
Comment (pshriwise) [inline pyne/openmc_utils.py:418]: It might be good to update the `MeshTally.__init__` to take some of these parameters like the `tally_id`, `particle`, `structured_coords`, etc. Right now we create a partially functional object and then fill in additional information as we go. That happens both here and in the `mcnp` module. It’d probably be better to have a fully functional object from the start by collecting all this information in the `create_meshtally` functions and then creating a `MeshTally` instance by passing all necessary information at once.

Could be left to a future PR though.

File: pyne/openmc_utils.py [hunk 10]
```diff
@@ -6,17 +6,33 @@
...
+         QAWarning)
 from pyne import nucname
 from pyne.utils import QAWarning
 
 warn(__name__ + " is not yet QA compliant.", QAWarning)
 
+try:
+    import openmc
+except:
+    warn("The openmc (OpenMC Python API) could not be imported. "
```
Comment (pshriwise) [inline pyne/openmc_utils.py:33]: Should this be more than a warning? Is there any remaining functionality in this module if we can’t import `OpenMC`?

**Review by pshriwise — CHANGES_REQUESTED** [review]
Quick initial review. This is awesome, @zxkjack123. Excited to see this work appearing in PyNE!

In addition to my line comments, this implementation currently supports exactly a `MeshFilter` tally scoring flux and I think expects that an `EnergyFilter` is also applied. This is fine to start, but we may want to explain why and check for those conditions on the tally in the state point file.

File: pyne/openmc_utils.py (hunk 3 above)
Comment (gonuke) [reply pyne/openmc_utils.py:348]: That just reverses it, but that's not what we need here, I don't think... but maybe something similar to...

```
results = np.transpose(results.reshape((x_dim,y_dim,z_dim)).reshape((1,x_dim*y_dim*z_dim))
```

or something like that?

**Review by gonuke — COMMENTED** [review]
I am going to leave this to @pshriwise since he's more familiar with the OpenMC formats.  Thanks for the contribution @zxkjack123

File: pyne/openmc_utils.py (hunk 3 above)
Comment (pshriwise) [reply pyne/openmc_utils.py:348]: Ah, yeah. Very true. Was a little confused there.

Comment (pshriwise) [reply pyne/openmc_utils.py:348]: @zxkjack123 I found the following works as a replacement for `result_changes_order` in the 2-D mesh examples from the test cases. There might need to be another `np.moveaxis` call if there is more than one element in the third dimensions - not sure. 

```python
    exp_result = np.moveaxis(exp_result.reshape(num_e_groups, *dims), 0, 1).reshape(num_ves, num_e_groups)
```

File: pyne/openmc_utils.py (hunk 2 above)
Comment (zxkjack123) [reply pyne/openmc_utils.py:235]: The `MeshFilter.mesh` check has been added in https://github.com/pyne/pyne/pull/1202/commits/666235792608970260e17ba595f4509dd6418add.

File: pyne/openmc_utils.py (hunk 3 above)
Comment (zxkjack123) [reply pyne/openmc_utils.py:348]: Thank you @gonuke , @pshriwise . The `np.moveaxis` is used simplify the order change of the result. See https://github.com/pyne/pyne/pull/1202/commits/134f1b730400262dd5b44b52d26bc0f1827b09e7.

File: pyne/openmc_utils.py (hunk 7 above)
Comment (zxkjack123) [reply pyne/openmc_utils.py]: It should report an error if `mesh_has_mats` is True, as it not yet supported. Updated in https://github.com/pyne/pyne/pull/1202/commits/d65a316f8b9d5f8782c9c280a76461bd3c70793d.

File: pyne/openmc_utils.py (hunk 9 above)
Comment (zxkjack123) [reply pyne/openmc_utils.py:418]: Agreed. That will be in a future PR.

File: pyne/openmc_utils.py (hunk 10 above)
Comment (zxkjack123) [reply pyne/openmc_utils.py:33]: Yes, there are. For example, the class `AceTable` is not dependent on OpenMC Python API.

File: pyne/openmc_utils.py (hunk 6 above)
Comment (zxkjack123) [reply pyne/openmc_utils.py:435]: The `x/y/z/e_bounds` are converted to `tuple` in https://github.com/pyne/pyne/pull/1202/commits/4e105228c96813dc750e61a9c9411bd4230dc00f.

File: pyne/mcnp.py [hunk 11]
```diff
@@ -2123,11 +2119,11 @@ def read_xyze_bounds(self, f):
         while line.strip() != 'Tally bin boundaries:':
             line = f.readline()
 
-        x_bounds = [float(x) for x in f.readline().split()[2:]]
-        y_bounds = [float(x) for x in f.readline().split()[2:]]
-        z_bounds = [float(x) for x in f.readline().split()[2:]]
+        x_bounds = tuple([float(x) for x in f.readline().split()[2:]])
```
Comment (pshriwise) [inline pyne/mcnp.py]: Small change for readability. Applies to `y/z/e_bounds` too.

```suggestion
        x_bounds = tuple(float(x) for x in f.readline().split()[2:])
```

File: pyne/openmc_utils.py [hunk 12]
```diff
@@ -156,3 +171,274 @@ def xml(self):
...
+    -----------
+    filename : str
+        The OpenMC state point file name.
+    tally_id : int
+        Tally id to read.
+
+    Returns:
+    --------
+    e_bounds : numpy array
+        Energy boundries with size of (num_e_gourps + 1).
```
Comment (pshriwise) [inline pyne/openmc_utils.py]:
```suggestion
        Energy boundries with size of (num_e_groups + 1).
```

File: pyne/openmc_utils.py [hunk 13]
```diff
@@ -156,3 +171,274 @@ def xml(self):
...
+    for flt in tally.filters:
+        if isinstance(flt, openmc.filter.EnergyFilter):
+            energy_filter = flt
+    e_bounds = energy_filter.values
+    return e_bounds
+
+
+def get_structured_coords_from_openmc_sp(filename, tally_id):
+    """
+    This function read the OpenMC state point file and get the structured
```
Comment (pshriwise) [inline pyne/openmc_utils.py]:
```suggestion
    This function reads the OpenMC state point file and gets the structured
```

File: pyne/openmc_utils.py [hunk 14]
```diff
@@ -156,3 +171,274 @@ def xml(self):
...
+
+    sp = openmc.StatePoint(filename)
+    tally = sp.get_tally(id=tally_id)
+    # check filters to find MeshFilter
+    for flt in tally.filters:
+        if isinstance(flt, openmc.filter.MeshFilter):
+            if isinstance(flt.mesh, openmc.RegularMesh):
+                mesh_filter = flt
+            else:
+                raise NotImplementedError("Current MeshFilter.mesh is "
```
Comment (pshriwise) [inline pyne/openmc_utils.py]: This is a nice change. Thank you @zxkjack123. I might update a bit to say something like: 

```python
raise NotImplementedError("The mesh on this tally: {} is not yet supported. Please use an openmc.RegularMesh".format(flt.mesh))
```

File: pyne/openmc_utils.py [hunk 15]
```diff
@@ -156,3 +171,274 @@ def xml(self):
...
+    structured_coords = calc_structured_coords(
+        mesh_filter.mesh.lower_left[:],
+        mesh_filter.mesh.upper_right[:],
+        mesh_filter.mesh.dimension[:])
+    return structured_coords
+
+
+def calc_structured_coords(lower_left, upper_right, dimension):
+    """
+    This function calculate the structured mesh coordinations from OpenMC mesh
```
Comment (pshriwise) [inline pyne/openmc_utils.py]:
```suggestion
    This function calculates the structured mesh coordinates from an OpenMC mesh
```

File: pyne/openmc_utils.py [hunk 16]
```diff
@@ -156,3 +171,274 @@ def xml(self):
...
+    parameters.
+
+    Parameters:
+    -----------
+    lower_left : numpy array of float
+        The lower left coordinate of the mesh. A numpy array of lenght 3.
```
Comment (pshriwise) [inline pyne/openmc_utils.py]:
```suggestion
        The lower left coordinate of the mesh. A numpy array of length 3.
```

File: pyne/openmc_utils.py [hunk 17]
```diff
@@ -156,3 +171,274 @@ def xml(self):
...
+        Format: [x_min, y_min, z_min].
+    upper_right : numpy array of float
+        The upper right coordinate of the mesh. A numpy array of lenght 3.
```
Comment (pshriwise) [inline pyne/openmc_utils.py]:
```suggestion
        The upper right coordinate of the mesh. A numpy array of length 3.
```

File: pyne/openmc_utils.py [hunk 18]
```diff
@@ -156,3 +171,274 @@ def xml(self):
...
+    structured_coords : numpy array
+        A nested numpy array definning the boundaries of the mesh element in
+        each dimension. Format: [[x_bounds1, x_bounds2. ...],
+                                 [y_bounds1, y_bounds2, ...],
+                                 [z_bounds1, z_bounds2, ...]]
+    """
+
+    # check the length of parameters
+    if len(lower_left) != 3 or len(upper_right) != 3 or len(dimension) != 3:
+        raise ValueError("Only 3D OpenMC mesh is supported!")
```
Comment (pshriwise) [inline pyne/openmc_utils.py]: Let's go ahead and raise a `NotImplementedError` here too.

File: pyne/openmc_utils.py [hunk 19]
```diff
@@ -156,3 +171,274 @@ def xml(self):
...
+    -----------
+    filename : str
+        The OpenMC state point file name.
+    m : MeshTally
+        MeshTally for the tally.
+
+    Returns:
+    --------
+    result : numpy array
+        This numpy array contains the flux data read from OpenMC state point
```
Comment (pshriwise) [inline pyne/openmc_utils.py]:
```suggestion
        This numpy array contains the flux data read from the OpenMC state point
```

File: pyne/openmc_utils.py [hunk 20]
```diff
@@ -156,3 +171,274 @@ def xml(self):
...
+        file. The shape of this numpy array is (num_ves*num_e_groups), with z
+        changes fastest.
+    rel_error: numpy array
+        This numpy array contains the relative error data.
+    res_tot : list
```
Comment (pshriwise) [inline pyne/openmc_utils.py]:
```suggestion
    res_tot : numpy array
```

File: pyne/openmc_utils.py [hunk 21]
```diff
@@ -156,3 +171,274 @@ def xml(self):
...
+        The total results.
+    rel_err_tot : list
```
Comment (pshriwise) [inline pyne/openmc_utils.py]:
```suggestion
    rel_err_tot : numpy array
```

File: pyne/openmc_utils.py [hunk 22]
```diff
@@ -156,3 +171,274 @@ def xml(self):
...
+        Relative error of total results.
+    """
+
+    sp = openmc.StatePoint(filename)
+    tally = sp.get_tally(scores=['flux'], id=m.tally_number)
+    flux = tally.get_slice(scores=['flux'])
+
+    num_ves = len(m)
+    # currently, the openmc mesh are uniform
```
Comment (pshriwise) [inline pyne/openmc_utils.py]:
```suggestion
    # assumption: openmc mesh voxels volumes are uniform for a regular mesh
```

File: pyne/openmc_utils.py (hunk 3 above)
Comment (pshriwise) [reply pyne/openmc_utils.py:348]: Excellent!

File: pyne/openmc_utils.py [hunk 23]
```diff
@@ -156,3 +171,274 @@ def xml(self):
...
+    result = result_changes_order(result, m.dims[3:6])
+    rel_err = result_changes_order(rel_err, m.dims[3:6])
+    res_tot = result_changes_order(res_tot, m.dims[3:6])
+    rel_err_tot = result_changes_order(rel_err_tot, m.dims[3:6])
+    return result, rel_err, res_tot, rel_err_tot
+
+def result_changes_order(result, dims):
+    """
+    This function changes the order of openmc flux data. The default order of
+    OpenMC flux is x changes the fastest. This function changes it to the order
```
Comment (pshriwise) [inline pyne/openmc_utils.py]:
```suggestion
    OpenMC flux is x changes the fastest. This function reorganizes the data such that z changes fastest (the same order as the MCNP meshtal file).
```

File: pyne/openmc_utils.py [hunk 24]
```diff
@@ -156,3 +171,274 @@ def xml(self):
...
+    which z changes the fasest (the same order as MCNP meshtal).
+
+    Parameters:
+    -----------
+    result : numpy array
+        This numpy array contains the flux data read from OpenMC state point
+        file. The shape of this numpy array is (num_ves*num_e_groups) or
```
Comment (pshriwise) [inline pyne/openmc_utils.py]:
```suggestion
        file. The shape of this numpy array is (num_ves, num_e_groups) or
```

File: pyne/openmc_utils.py [hunk 25]
```diff
@@ -156,3 +171,274 @@ def xml(self):
...
+        (num_ves, ), with x changes the fastest.
```
Comment (pshriwise) [inline pyne/openmc_utils.py]:
```suggestion
        (num_ves, ), with x changing the fastest.
```

File: pyne/openmc_utils.py [hunk 26]
```diff
@@ -156,3 +171,274 @@ def xml(self):
...
+    dims : list
+        The dimension of x, y, z.
+
+    Returns:
+    --------
+    result_ : numpy array
+        This numpy array contains the flux data read from OpenMC state point
+        file. The shape of this numpy array is (num_ves*num_e_groups) or
```
Comment (pshriwise) [inline pyne/openmc_utils.py]:
```suggestion
        file. The shape of this numpy array is (num_ves, num_e_groups) or
```

File: pyne/openmc_utils.py [hunk 27]
```diff
@@ -156,3 +171,274 @@ def xml(self):
...
+        (num_ves, ), with changes the fastest.
```
Comment (pshriwise) [inline pyne/openmc_utils.py]:
```suggestion
        (num_ves, ), with z chaning the fastest.
```

File: pyne/openmc_utils.py [hunk 28]
```diff
@@ -156,3 +171,274 @@ def xml(self):
...
+    """
+
+    x_dim, y_dim, z_dim = dims[0], dims[1], dims[2]
+    num_e_groups = len(result.flatten()) // (x_dim * y_dim * z_dim)
```
Comment (pshriwise) [inline pyne/openmc_utils.py]:
```suggestion
    num_e_groups = result.size // (x_dim * y_dim * z_dim)
```

File: pyne/openmc_utils.py [hunk 29]
```diff
@@ -156,3 +171,274 @@ def xml(self):
...
+    result_ = np.reshape(result, newshape=(z_dim, y_dim, x_dim, num_e_groups))
+    # set energy changes fastest, z next, y next and then x
+    result_ = np.moveaxis(result_, [0, 1, 2, 3], [2, 1, 0, 3])
+    result_ = np.reshape(result_, newshape=(x_dim*y_dim*z_dim, num_e_groups))
```
Comment (pshriwise) [inline pyne/openmc_utils.py]:
```suggestion
    result_ = np.reshape(result_, newshape=(x_dim * y_dim * z_dim, num_e_groups))
```

File: pyne/openmc_utils.py [hunk 30]
```diff
@@ -156,3 +171,274 @@ def xml(self):
...
+    result_ : numpy array
+        This numpy array contains the flux data read from OpenMC state point
+        file. The shape of this numpy array is (num_ves*num_e_groups) or
+        (num_ves, ), with changes the fastest.
+    """
+
+    x_dim, y_dim, z_dim = dims[0], dims[1], dims[2]
+    num_e_groups = len(result.flatten()) // (x_dim * y_dim * z_dim)
+    result_ = np.reshape(result, newshape=(z_dim, y_dim, x_dim, num_e_groups))
+    # set energy changes fastest, z next, y next and then x
```
Comment (pshriwise) [inline pyne/openmc_utils.py]: Can you include some info about which is the axis for energy bounds here? Or just generally explain the axes transformation here in more detail? 

That was the tricky part to figure out (for me). And I think that someone coming upon this line of code later will have a lot of trouble understanding this function without it.

File: pyne/openmc_utils.py [hunk 31]
```diff
@@ -156,3 +171,274 @@ def xml(self):
...
+    result_ = np.moveaxis(result_, [0, 1, 2, 3], [2, 1, 0, 3])
+    result_ = np.reshape(result_, newshape=(x_dim*y_dim*z_dim, num_e_groups))
+    return result_
+
+
+def create_meshtally(filename, tally_id, mesh_id=None, particle=None,
+                     tag_names=None, mesh_has_mats=False):
+    """
+    This function creates a MeshTally instance from OpenMC statepoint file.
```
Comment (pshriwise) [inline pyne/openmc_utils.py]:
```suggestion
    This function creates a MeshTally instance from OpenMC state point file.
```

File: pyne/openmc_utils.py [hunk 32]
```diff
@@ -156,3 +171,274 @@ def xml(self):
...
+
+    Parameters:
+    -----------
+    filename : str
+        Filename of the OpenMC statepoint file. It ends with ".h5",
+        eg: "statepoint.10.h5".
+    tally_id : int
+        Tally number.
+    mesh_id : int
```
Comment (pshriwise) [inline pyne/openmc_utils.py]: Why is this required if there are multiple meshes? We should be able to look it up on the tally of interest right?

File: pyne/openmc_utils.py [hunk 33]
```diff
@@ -156,3 +171,274 @@ def xml(self):
...
+        Mesh id of this tally used. Required if multiple meshes exist in the
+        OpenMC state point file.
+    particle : str
+        The particle type, 'neutron' or 'photon'.
+    tag_names : iterable, optional
+        Four strs that specify the tag names for the results, relative
+        errors, total results and relative errors of the total results.
+    mesh_has_mats: bool
```
Comment (pshriwise) [inline pyne/openmc_utils.py:407]: Given that we throw an error if this is set to `True`, should we even have it as an option here?

File: pyne/openmc_utils.py [hunk 34]
```diff
@@ -156,3 +171,274 @@ def xml(self):
...
+    else:
+        m.tag_names = tag_names
+    # parameters to create mesh
+    structured_coords = get_structured_coords_from_openmc_sp(
+        filename, tally_id=m.tally_number)
+    m.x_bounds = tuple(structured_coords[0])
+    m.y_bounds = tuple(structured_coords[1])
+    m.z_bounds = tuple(structured_coords[2])
+    m.dims = [len(m.x_bounds) - 1, len(m.y_bounds) - 1, len(m.z_bounds) - 1]
+    m.num_ves = (len(m.x_bounds)-1) * (len(m.y_bounds)-1) * (len(m.z_bounds)-1)
```
Comment (pshriwise) [inline pyne/openmc_utils.py]:
```suggestion
    m.num_ves = np.prod(m.dims)
```

File: pyne/openmc_utils.py [hunk 35]
```diff
@@ -156,3 +171,274 @@ def xml(self):
...
+    m.e_bounds = tuple(get_e_bounds_from_openmc_sp(filename, m.tally_number))
+    m.num_e_groups = len(m.e_bounds) - 1
+    if mesh_has_mats:
```
Comment (pshriwise) [inline pyne/openmc_utils.py:439]: see comment above about this

File: tests/test_mcnp.py [hunk 36]
```diff
@@ -1034,16 +1034,16 @@ def test_single_meshtally_meshtal():
     assert_equal(meshtal_object.tally[4].dose_response, True)
     assert_equal(
         meshtal_object.tally[4].x_bounds,
-        [-200.00, -66.67, 66.67, 200.00])
+        (-200.00, -66.67, 66.67, 200.00))
```
Comment (pshriwise) [inline tests/test_mcnp.py:1037]: Good adjustment. :+1:

File: tests/test_openmc_utils.py [hunk 37]
```diff
@@ -0,0 +1,274 @@
...
+    # energy bin: [0.0, 1.0, 20.0], 2bins
+    # 6 volume elemenes
+    filename = os.path.join(cwd, "files_test_openmc", "statepoint.10.ebin2.ves6.h5")
+    # OpenMC energy unit is eV
+    exp_e_bounds = np.array([0.0, 1.0, 20.0]) * 1e6
+    e_bounds = openmc_utils.get_e_bounds_from_openmc_sp(filename, tally_id=1)
+    assert_array_equal(e_bounds, exp_e_bounds)
+
+def test_flux_changes_order():
+    if not HAVE_PYMOAB or sys.version_info[0] == 2:
```
Comment (pshriwise) [inline tests/test_openmc_utils.py]: This test can be run if we don't have `openmc` or `pymoab`

File: pyne/openmc_utils.py [hunk 38]
```diff
@@ -156,3 +171,274 @@ def xml(self):
...
+    OpenMC flux is x changes the fastest. This function changes it to the order
+    which z changes the fasest (the same order as MCNP meshtal).
+
+    Parameters:
+    -----------
+    result : numpy array
+        This numpy array contains the flux data read from OpenMC state point
+        file. The shape of this numpy array is (num_ves*num_e_groups) or
+        (num_ves, ), with x changes the fastest.
+    dims : list
```
Comment (pshriwise) [inline pyne/openmc_utils.py:361]: tuple?

**Review by pshriwise — CHANGES_REQUESTED** [review]
Some more comments. Nothing major on design.

File: pyne/openmc_utils.py (hunk 14 above)
Comment (zxkjack123) [reply pyne/openmc_utils.py]: Update in https://github.com/pyne/pyne/pull/1202/commits/5be5672c550b6e3fc3d9472b9fb5a33be05f0ad5.

File: pyne/openmc_utils.py (hunk 18 above)
Comment (zxkjack123) [reply pyne/openmc_utils.py]: Applied in https://github.com/pyne/pyne/pull/1202/commits/ecbbae05cf854db7fb6d00146e64e48416ec405b.

File: pyne/openmc_utils.py (hunk 30 above)
Comment (zxkjack123) [reply pyne/openmc_utils.py]: More detail description about this part is added in https://github.com/pyne/pyne/pull/1202/commits/89cff448ae2eb75d234323074630cd36d7f3c053.

File: pyne/openmc_utils.py (hunk 32 above)
Comment (zxkjack123) [reply pyne/openmc_utils.py]: Nice catch. In the very beginning version of this work, I used `mesh_id` to find out the mesh because OpenMC Python API is not yet an optional dependency (I can't differentiate mesh filter and energy filter) .

But later on, when OpenMC Python API was added as an optional dependency, I sued the filter check to find out the corresponding mesh filter for a given tally. Therefore, the `mesh_id` is not required now. But I forgot to remove it.

Thanks for notice this, and the `mesh_id` is remove in https://github.com/pyne/pyne/pull/1202/commits/5670c4d7e6fb1dc4a7b5af0339a41045b07e6be4.

File: pyne/openmc_utils.py (hunk 33 above)
Comment (zxkjack123) [reply pyne/openmc_utils.py:407]: I suggest to kept it here reminding us to implement it. I have changed the error message about it from `ValueError` to `NotImplementedError`.

File: tests/test_openmc_utils.py (hunk 37 above)
Comment (zxkjack123) [reply tests/test_openmc_utils.py]: You are right. Updated in https://github.com/pyne/pyne/pull/1202/commits/88003563c7db8e812f5aa5749c7f09dcf9b63c2b.

File: pyne/openmc_utils.py (hunk 38 above)
Comment (zxkjack123) [reply pyne/openmc_utils.py:361]: I found that in class `Mesh`, the `dims` and boundaries are stored as `list` or `list of list`. It's better to make a separate PR to change them to `tuple` at once.

Comment (pshriwise) [reply pyne/openmc_utils.py:361]: Sounds good to me.

**Comment by zxkjack123** [comment]
@pshriwise Any further comment/suggestion for this PR?

File: pyne/openmc_utils.py (hunk 33 above)
Comment (pshriwise) [reply pyne/openmc_utils.py:407]: I'm curious though, in what scenario will we have materials coming from a mesh in an OpenMC state point file?

**Review by pshriwise — COMMENTED** [review]
Just one comment to resolve and I then I think this is ready.

**Comment by zxkjack123** [comment]
> Just one comment to resolve and I then I think this is ready.

@pshriwise The materials are not from OpenMC state point file. The docstring about this in `class Mesh` says:
```
mats : MaterialLibrary or dict or Materials or None, optional
            This is a mapping of volume element handles to Material objects.
            If mats is None, then no empty materials are created for the mesh.

            Unstructured mesh instantiation:
                 - From PyMOAB core instance by specifying: <mesh>
                 - From mesh file by specifying: <mesh_file>

            Structured mesh instantiation:
                - From PyMOAB core instance with exactly 1 entity set
                  (with BOX_DIMS tag) by specifying <mesh> and structured = True.
                - From mesh file with exactly 1 entity set (with BOX_DIMS tag)
                  by specifying <mesh_file> and structured = True.
                - From a PyMOAB instance with multiple entity sets by
                  specifying <mesh>, <structured_set>, structured=True.
                - From coordinates by specifying <structured_coords>,
                  structured=True, and optional pre-existing PyMOAB core
                  instance <mesh>
```
The materials are come from a PyMOAB core instance or a `h5` mesh file. I have never created a meshtally with material, not sure about the format of these materials.

**Comment by pshriwise** [comment]
I see, I was misinterpreting `mesh_has_mats` as meaning that it would try to read materials from the mesh.

**Review by pshriwise — APPROVED** [review]
Looks good @zxkjack123!
